subprojects/docs/src/docs/userguide/java.adoc

```diff
@@ -0,0 +1,453 @@
+
+[[sec:running_java_tests]]
+=== Running tests
+
+Alongside providing automatic compilation of unit tests in _src/test/java_, the Java Plugin has native support for running tests that use JUnit 3, 4 & 5 (JUnit 5 support https://docs.gradle.org/4.6/release-notes.html#junit-5-support[came in Gradle 4.6]) and TestNG. You get:
+
+ * An automatic `test` task of type api:org.gradle.api.tasks.testing.Test[], using the `test` source set
+ * An HTML test report that includes the results from _all_ `Test` tasks that run
+ * Easy filtering of which tests to run
+ * Fine-grained control over how the tests are run
+ * The opportunity to create your own test execution and test reporting tasks
+
+You do _not_ get a `Test` task for every source set you declare, since not every source set represents tests! That's why you typically need to <<sec:custom_java_source_sets,create your own `Test` tasks>> for things like integration and acceptance tests if they can't be included with the `test` source set.
+
+As there is a lot to cover when it comes to testing, the topic has its <<java_testing,own chapter>> in which we look at:
+
+ * How tests are run
+ * How to run a subset of tests via filtering
+ * How Gradle discovers tests
+ * How to configure test reporting and add your own reporting tasks
+ * How to make use of specific JUnit and TestNG features
+
+You can also learn more about configuring tests in the DSL reference for api:org.gradle.api.tasks.testing.Test[].
+
+
+[[sec:java_packaging]]
+=== Packaging and publishing
+
+How you package and potentially publish your Java project depends on what type of project it is. Libraries, applications, web applications and enterprise applications all have differing requirements. In this section, we will focus on the bare bones provided by the Java Plugin.
+
+The one and only packaging feature provided by the Java Plugin directly is a `jar` task that packages all the compiled production classes and resources into a single JAR. This JAR is then added as an artifact — as opposed to a dependency — in the `archives` configuration, hence why it is automatically built by the `assemble` task.
+
+If you want any other JAR or alternative archive built, you either have to apply an appropriate plugin or create the task manually. For example, if you want a task that generates a 'sources' JAR, define your own `Jar` task like so:
+
+++++
+<sample id="definingSourcesJarTask" dir="userguide/java/basic" title="Defining a custom task to create a 'sources' JAR">
+    <sourcefile file="build.gradle" snippet="defining-sources-jar-task"/>
+</sample>
+++++
+
+See api:org.gradle.api.tasks.bundling.Jar[] for more details on the configuration options available to you.
+
+There are several options for publishing a JAR once it has been created, such as the `uploadArchives` task that can publish to Maven- and Ivy-compatible repositories. It's a big topic, so please see _<<artifact_management,Publishing artifacts>>_ for more details.
```

Comment:
`artifact_management` is only the old publishing mechanism. We're really only covering publishing in the very last paragraph. It seems like this would be worth expanding and linking off to the 3 relevant chapters, or omitting publishing here.

Reply:
I'm not sure what we can say about the two. I still don't know which is the preferred approach, if any. Is there anything useful we can say?

Reply:
OK, so @oehme says that this will be resolved with Gradle 4.8: the old mechanism will be deprecated. So I suggest mentioning all publishing options here without any major guidance, recognising that we will be updating this section (and others I suspect) with the release of the next version of Gradle.